### ISSUE TITLE
[A11Y FEATURE]: The page <title> should not be empty

### ISSUE THREAD

**Issue by alvaromontoro** [jackdomleo7/Checka11y.css#153]
# A11y Feature Request

## Describe the new a11y feature

Web pages have titles that describe their topic or purpose. Titles go in the `<title>` tag inside the `<head>`, and are displayed on the top bar of the browser as well as announced by assistive technologies (AT).

The new feature would be to show an error/warning if the title is missing or empty. This will match the [WCAG Level A accessibility criterion 2.4.2](https://www.w3.org/WAI/WCAG21/quickref/?versions=2.0&showtechniques=242#qr-navigation-mechanisms-title).

## Acceptance criteria

- [ ] The `<title>` tag must be present in the `<head>`.
- [ ] The `<title>` tag must have descriptive content. 

### Possible solution

Detecting and showing an error if the `<title>` tag is missing is going to be complicated... but, we can cover part of the criteria: an empty title. We can use the `:empty` pseudo-class applied to the title to show a message if the `<title>` has no content (which means that the title is not descriptive.)

Here is a _simple_ approach that will display an error (should it be a warning instead?) when a page has an empty title:

```css
/* Make the head visible to show the title. This could have side effects but,
in theory, all the elements in the head are non-visible. So we should be ok :P */
head {
  display: block;
}

/* Special case, remove the border and add display: block for it to be visible. */
head title:empty {
  @include vectorMessage(error, "XXXXX", 458, "the title tag should not", "be empty as per criterion 2.4.2.");
  display: block;
  border: 0;
}
```

I tested it on MacOS with Chrome, Firefox, and Safari, and they all show the error message if the `<title>` is empty.

## Why is this a11y feature important?

A missing or empty title would be an accessibility issue because it makes it more difficult for some users to orient themselves and navigate through the pages (see [WCAG techniques for more information](https://www.w3.org/WAI/WCAG21/Techniques/general/G88.html))

**Comment by alvaromontoro** [comment]
I don't think the first acceptance criteria is possible to match only with CSS (it would be if browsers supported [`:has() relational pseudo-class`](https://developer.mozilla.org/en-US/docs/Web/CSS/:has)... but they don't 😢).

Maybe we could add a rule for that. No browser will match it at the moment, but it could work (one can hope) in the future. Something like this:

```css
head:not(:has(title)) {
  @include vectorMessage(error, "XXXXX", 450, "a title tag must be present", "in the page head.");
  display: block;
  border: 0;
}
```

**Comment by jackdomleo7** [comment]
Good idea @alvaromontoro! I'm happy to go ahead with this feature!

For the criterion where the `<title>` isn't present, I'm happy with your suggested approach as long as it doesn't impact browsers that don't support it. I'd also suggest in the Checka11y.css docs, we'd label it as "partial feature support" or something.

Thank you for suggesting!

**Comment by alvaromontoro** [comment]
Thank you, @jackdomleo7! Please assign the task to me :)

**Comment by alvaromontoro** [comment]
Exciting CSS news: https://twitter.com/jensimmons/status/1473051429115940868

This selector may end up working sooner than expected (at least on Safari):

```css
head:not(:has(title)) {
  @include vectorMessage(error, "XXXXX", 450, "a title tag must be present", "in the page head.");
  display: block;
  border: 0;
}
```

**Comment by jackdomleo7** [comment]
That's interesting!!
